### ISSUE TITLE
Unable to access any parameter gradients thru the Python interface during FP16 training

### ISSUE THREAD

**Issue by jarvis-huang** [NVIDIA/caffe#348]
Hi, I'm trying to perform FP16 training (storage=FP16, math=FP32) using AlexNet on CIFAR10 dataset. During training, I compared loss between the normal FP32 version and FP16 version and found that FP16 quickly diverged at around 400 iterations but FP32 could continue to train with decreasing loss. When I tried to inspect the gradients of "conv1" layer of the FP16, by 
`print np.max(net16.params['conv1'][0].diff), np.min(net16.params['conv1'][0].diff)`
immediate I got this error:
**F0607 06:31:43.975522 25044 tensor.cpp:57] Check failed: !locked_ Tensor is locked and its type FLOAT16 can't be changed to FLOAT**

All I did was adding
```
 default_forward_type:  FLOAT16
 default_backward_type: FLOAT16
 default_forward_math:  FLOAT
 default_backward_math: FLOAT
```
to the begining of AlexNet prototxt.

Training script is as follows:
```python
caffe_root = '/home/xxx/caffe/'
import sys
sys.path.insert(0, caffe_root + 'python')
import caffe
import numpy as np

caffe.set_mode_gpu()
caffe.set_device(1)

solver16 = caffe.SGDSolver('/xxx/solver_fp16.prototxt')
solver32 = caffe.SGDSolver('/xxx/solver_fp32.prototxt')
net16 = solver16.net
net32 = solver32.net

solver16.step(240)
solver32.step(240)

print np.max(net16.params['conv1'][0].diff), np.min(net16.params['conv1'][0].diff)
```

Is this a bug or something I did wrong here? Thanks for any help.

**Comment by drnikolaev** [comment]
hi @jarvis-huang thanks for reporting this. 
What particular solver_fp16.prototxt and solver_fp32.prototxt do you use? Could you please attach them here?

**Comment by jarvis-huang** [comment]
Here's the **solver_fp16.prototxt**. **solver_fp32.prototxt** looks almost the same but with different net and snapshot_prefix.

```bash
# The train/test net protocol buffer definition
net: "examples/cifar10/jhuang_fp16.prototxt"
# test_iter specifies how many forward passes the test should carry out.
# In the case of CIFAR10, we have test batch size 100 and 100 test iterations,
# covering the full 10,000 testing images.
test_iter: 100
# Carry out testing every 1000 training iterations.
test_interval: 1000
# The base learning rate, momentum and the weight decay of the network.
base_lr: 0.005
momentum: 0.9
weight_decay: 0.004
# The learning rate policy
lr_policy: "fixed"
# Display every 200 iterations
display: 200
# The maximum number of iterations
max_iter: 60000

snapshot: 10000
snapshot_prefix: "examples/cifar10/jhuang_fp16"
solver_mode: GPU
```

Programmatically, the error was caused by `lock_tensor()` from **tensor.hpp** and which is in turn called by 
`learnable_params_[i]->lock_diff()` inside **net.cpp**.  That explains why I don't get errors when I try to print gradients on layer outputs, but only on layer params.

I also tried running another `backward()` before printing gradients since somewhere I read says gradients were cleared after each `step()`, but the same error was triggered.

**Comment by drnikolaev** [comment]
Hi @jarvis-huang  Locking issue has been fixed in v0.16.3 release. Please try it.
